### PR TITLE
Implementación de pago con Bitcoins

### PR DESCRIPTION
**Pull request by fedebalderas**
Se realizo la implementación de pago con Bitcoins a través del Gateway de pago Openpay

### CODE REVIEW

File: openpayprestashop.php [hunk 1]
```diff
@@ -24,985 +24,1068 @@
...
-	}
-
-	public function createWebhook()
-	{
-		$domain = (Configuration::get('PS_SSL_ENABLED') ? 'https' : 'http').'://'.$_SERVER['HTTP_HOST'];
-		$webhook_data = array(
-			'url' => $domain.__PS_BASE_URI__.'modules/openpayprestashop/notification.php',
-			'user' => Configuration::get('OPENPAY_WEBHOOK_USER'),
-			'password' => Configuration::get('OPENPAY_WEBHOOK_PASSWORD'),
-			'event_types' => ar
```
Comment (Mekki78) [inline openpayprestashop.php]: download

PHP Fatal error:  Call to a member function create() on a non-object in openpaypresta
shop.php on line 897

(btw i am logged in)

**Comment by Quetzacoalt91** [comment]
Hi @fedebalderas,

In this new version, the configuration data and the database structure has changed.
You have to think to add these changes when the module is already installed !

To do so, you have to add an upgrade file for version 1.6.3. Here is a documentation to help you: http://doc.prestashop.com/display/PS15/Auto-updating+modules

File: views/templates/front/bitcoin_execution.tpl [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            <button type="submit"  class="button btn btn-default standard-checkout button-medium">
+                <span>
+                    {l s='Generate payment address' mod='openpayprestashop'}
+                    <i class="icon-chevron-right right"></i>
+                </span>
+            </button>
+        </p>
+    </form>
+
+    <script type="text/javascript">
```
Comment (Mekki78) [inline views/templates/front/bitcoin_execution.tpl]: Any reason of this empty script ?

File: controllers/front/bitcoinpayment.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        $id = Configuration::get('OPENPAY_MODE') ? Configuration::get('OPENPAY_MERCHANT_ID_LIVE') : Configuration::get('OPENPAY_MERCHANT_ID_TEST');
+
+        if (!empty($this->context->cookie->openpay_error)) {
+            $this->context->smarty->assign('openpay_error', $this->context->cookie->openpay_error);
+            $this->context->cookie->__set('openpay_error', null);
+        }
+
+        $this->context->smarty->assign(array(
+            'validation_url' => './index.php?process=validation&fc=module&module=openpayprestashop&controller=default',
+            'id' => $id,
```
Comment (Mekki78) [inline controllers/front/bitcoinpayment.php]: {$id} is not in the TPL file.

File: controllers/front/bitcoinpayment.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            'mode' => Configuration::get('OPENPAY_MODE'),
```
Comment (Mekki78) [inline controllers/front/bitcoinpayment.php]: {$mode} is not in the TPL file.

File: controllers/front/bitcoinpayment.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            'nbProducts' => $cart->nbProducts(),
+            'total' => $cart->getOrderTotal(true, Cart::BOTH),
```
Comment (Mekki78) [inline controllers/front/bitcoinpayment.php]: {$total} is not in the TPL file.

File: openpayprestashop.php (hunk 1 above)
Comment (fedebalderas) [inline openpayprestashop.php]: It's looks like if Openpay customer wasn't created.
Can i ask you a favor, could you take a screenshot of your "login Openpay account? Just to corroborate this theory and check why wasn't created.

Thanks!

File: views/templates/front/bitcoin_execution.tpl (hunk 2 above)
Comment (fedebalderas) [inline views/templates/front/bitcoin_execution.tpl]: This issue was solved and will be updated when issue on line 897 is resolved.

File: controllers/front/bitcoinpayment.php (hunk 3 above)
Comment (fedebalderas) [inline controllers/front/bitcoinpayment.php]: This issue was solved and will be updated when issue on line 897 is resolved

File: controllers/front/bitcoinpayment.php (hunk 4 above)
Comment (fedebalderas) [inline controllers/front/bitcoinpayment.php]: This issue was solved and will be updated when issue on line 897 is resolved

File: controllers/front/bitcoinpayment.php (hunk 5 above)
Comment (fedebalderas) [inline controllers/front/bitcoinpayment.php]: This issue was solved and will be updated when issue on line 897 is resolved

**Comment by fedebalderas** [comment]
@Quetzacoalt91 i'm a little confused about upgrade file name, examples use 2: Upgrade-1.2.php and install-1.2.php. Which is correct?

**Comment by Quetzacoalt91** [comment]
To test it, you have to install the existing version from PrestaShop Addons and then upload a ZIP file with your upgraded module. You will see a message that a upgrade file has been successfully applied.

Best regards

**Comment by fedebalderas** [comment]
@Quetzacoalt91 Upgrade implementation has done
modules farma listo 2015-09-24 18-43-33

Thomas can i ask a favor, could you upgrade your Openpay's module to verify that is correct, and besides can you simulate a Bitcoin payment in your local store.

I think that the problem that @Mekki78  reported has relations with differences between version 1.6.2 and 1.6.3, maybe with the upgrade implementation this problem is solved.

**Comment by Quetzacoalt91** [comment]
Hi @fedebalderas,

Unfortunately, I get the same error as @Mekki78.
```
Notice: Trying to get property of non-object in /home/modules/www/public_html/thomas/modules/openpayprestashop/openpayprestashop.php on line 897

Fatal error: Call to a member function create() on null in /home/modules/www/public_html/thomas/modules/openpayprestashop/openpayprestashop.php on line 897
```

But the upgrade file was properly executed ! :+1:

**Comment by fedebalderas** [comment]
I'm trying to repeat this error and I can't, this error is just witch Bitcoin payment? 
Can i ask you another favor?  could you check your log in Openpay's dashboard just to identify what happen here.

Thanks Thomas and sorry bother you

**Comment by Quetzacoalt91** [comment]
This error also happen when I select another payment mode of OpenPay. I will have a look on the dashboard

**Comment by Quetzacoalt91** [comment]
Actually, we do not have access to our OpenPay dashboard. We only have API keys to use in the module.

However, when I save the credentials in the module configuration, here is the answer I get:
capture du 2015-09-25 15 57 51

**Comment by fedebalderas** [comment]
@Quetzacoalt91 @Mekki78 We already found the error, I tested your Test's Store (http://modules-thomas.test.prestashop.com) and we notice that customer's address form doesn't show State field (in our address form it does), so then when we create a Openpay customer we send all address parameters and State's parameter is null and Openpay customer is never created.

So we just validate that all address parameters are NOT NULL.

openpay dashboard
address prestashop

**Comment by Mekki78** [comment]
Hi, 

**CRITICAL**
You must create your order ONLY after the payment has been completed.
After this step, you can redirect your customer on the Confirmation Page.
You cannot make the payment on the confirmation page, you need to follow the process you made for your other payment methods.

The cart will be validated even if he didn't try to pay. The customer will not be able to go back or choose an other payment method because PrestaShop will say that his cart is now empty.

download 4


Thank you,
Regards,

**Comment by fedebalderas** [comment]
@Mekki78 you mean when customer make a payment by credit card?

**Comment by Quetzacoalt91** [comment]
Hi @fedebalderas,

The point here is to create an "order" in PrestaShop only when the customer has completed his Bitcoin payment on your interface.

When Melvin talked about credit card, he was giving you an example:
* The customer reaches the payment providers list
* The customer selects "Bitcoin"
* An order is created (which is wrong) and the customer is redirected to the payment page
* The customer sees that he has selected the wrong payment mode and goes back to the payment providers
* The customer cannot select credit card (in example) because an order has been already created and the shopping cart has been emptied.

Do you now see what we mean ?

Best regards

**Comment by fedebalderas** [comment]
With Bitcoin, the order is created until the user clic on "Generate payment address", before that click, user can choose another payment method. In this step Order is NOT generated yet.
Thats why i'm a little confused with Melvin's comment.

step1
step2
step3
step4

**Comment by Mekki78** [comment]
Hello,

When your are on the Step 3, try to go back on your step 1 you will have now "Your shopping cart is empty"  even if you are not making the payment.

You should make the payment at the Step 2 and not at the Step 3 if that can be more clear for you. At this moment you are making the payment on the Order-Confirmation Page where you should normally have the order number and a confirmation message.

**Comment by fedebalderas** [comment]
That's right, when you are on step 3 there no way back... Openpay transaction is already generated and PrestaShop Order are too.
Step 2 is informative and for confirmation, in other way, users can choose another payment method.
But that's the business logic that Openpay transmitted to me and i wrote the code in that way.

Maybe you need to talk with IT Manager and expose your logic.

**Comment by Quetzacoalt91** [comment]
Hi @fedebalderas,

I'm sure we can find a solution for this issue. Unfortunately, the Order (on PrestaShop) must be only created when the customer reaches the customer gets the confirmation after the step 4.

Because actually, when he tries to pay by Bitcoins, he still can go back in order to select another payment provider.

If you really need to send an ID for the payment, you can send the cart ID. And when the payment will be completed, you will be able to create the Order with the Cart ID you just used.

Best regards

**Comment by fedebalderas** [comment]
Hi @Quetzacoalt91,
So then the issue is only with Bitcoin payment? If i understand, we need to generate PS Order when customer already paid, other way, customer can choose another payment method?

Best regards

**Comment by fedebalderas** [comment]
Hi @Quetzacoalt91,
In Bitcoin payment, to generate PS Order when customer already paid i will need to use the webhook to know that the payment was generated and create the Order, aja?

But how do i get: $this->context->cart->id or $this->currentOrder
```
/* Create the PrestaShop order in database */
        $this->validateOrder(
            (int) $this->context->cart->id,
            (int) $order_status,
            $result_json->amount,
            $display_name,
            $message,
            null,
            null,
            false,
            $this->context->customer->secure_key
        );
/** @since 1.5.0 Attach the Openpay Transaction ID to this Order */
        $new_order = new Order((int) $this->currentOrder);
        if (Validate::isLoadedObject($new_order)) {
            $payment = $new_order->getOrderPaymentCollection();
            if (isset($payment[0])) {
                $payment[0]->transaction_id = pSQL($result_json->id);
                $payment[0]->save();
            }
        }
```
Any idea?

**Comment by Quetzacoalt91** [comment]
Hi @fedebalderas,

Yes, the issue we talk about is only for Bitcoin payment.
For your two missing variables, you should send at least the cart id to the webhook. With this information, your can generate your Cart object, find the related Customer, and finally get its secure_key :)

**Comment by fedebalderas** [comment]
Hi @Quetzacoalt91,
I've just pushing some changes to the repository. Now the PS order is generated when customer made his bitcoin payment, i'm using Copay app to simulate the payment.

Best regards!!!

farma listo 2015-10-12 18-26-31

**Comment by Quetzacoalt91** [comment]
That great @fedebalderas, thank you.
I will ask for a security review of this improvement, just to be sure that everything is perfect. :wink:

**Comment by Mekki78** [comment]
Hello @fedebalderas,

I saw two new error in your module, the first one is related to the currency (see screenshot below this line)

download 8

And the second one is that i cannot be able to make the payment (maybe it's related).

12168134_534178540062908_1996885679_n

Thank you to correct this two issue, 
Regards,

**Comment by fedebalderas** [comment]
hi @Quetzacoalt91 
- In fact $json->transaction->id  is a string, so i just add pSQL (i will push this later)

hi @Mekki78 
- About second error is because you are using  a bitcoin app in production mode (real), to test bitcoin payment in sandbox mode, you need to use a test wallet (Testnet3)

http://www.openpay.mx/docs/bitcoin-charge.html
Android: Testnet3 
1.- Buscar en el play store la app "Testne3" e instalar.
2.- Al abrir la app, automáticamente ya contarán con un monedero que usa la red de pruebas "testnet" de bitcoins.

And about currencies error, i try to find a good way to restrict Openpay module to certain currencies, but i got nothing, do you have an official documentation about this?

Best regards!!!

**Comment by Quetzacoalt91** [comment]
Hi @fedebalderas,

For the currencies issue, here is a solution you could use:
* Add a variable inside your OpenPay module class
```php
public $limited_currencies = array('MXN');
```

* In your hook for the payment page, just check that you accept the currency:
```php
    public function hookPayment()
    {
        if (!$this->active) {
            return;
        }

        if (!in_array($this->context->currency->iso_code, $this->limited_currencies))
            return;
        }

    [...]
    }
```
If the currency is not supported, you can returndirectly (and nothing will be shown by the module). You can adapt the test in order to make this test only for bitcoins.

**Comment by fedebalderas** [comment]
hi @Quetzacoalt91 and @Mekki78
I'm trying to test the currencies issues, but when i select another currency that is not MXN, the payment methods disappears.
And if i'm in the payment confirm page and i choose another currency, i get redirect to the confirm Order page. So, i think that PrestaShop validate this issue internally.

Attached video:
https://youtu.be/zu7rzeFin7M

I already pushed the changes that @Quetzacoalt91 told me, about notification.php file (pSQL).

**Comment by Quetzacoalt91** [comment]
Hi @fedebalderas,

On my side your module is module is shown for every currency. That's why I recommend you to add this check.
(In your shop, the module does not appear thanks to your configuration in "*Modules & Services > Payment*")

**Comment by fedebalderas** [comment]
I didin´t know about the Payment configuration in  "Modules & Services > Payment.
I  already pushed the changes and i updated the version number.
I hope everything is OK :+1:

**Comment by johnz0r** [comment]
security review OK

**Comment by Quetzacoalt91** [comment]
Thanks for your quick review @johnz0r.

@fedebalderas, your PR just been merged and the changes will be available soon. :)

Best regards

**Comment by fedebalderas** [comment]
Thank you all @Quetzacoalt91 , @Mekki78 , @johnz0r 

Best regards!

**Pull request merged**